app/models/intake.rb

```diff
@@ -511,19 +511,20 @@ def determine_zendesk_instance_domain
 
   private
 
+  ##
+  # at load time, maps the source codes to the appropriate associated
+  # group id for speedy lookup
+  SOURCE_MATCHERS = Hash[SourceCode.all.map { |s| [s.code, s.vita_partner.zendesk_group_id] }]
```

Comment:
Caching this in a constant is a bit of an unnecessary optimization, IMHO, because of how infrequently we actually invoke the routing logic. That said, I think it's unlikely to cause confusion because the constant would be reloaded properly in development.

Also: We could implement longest prefix matching by sorting the SourceCode query by the code's length. (E.g. route `RefundDay-C` to Catalyst despite the `RefundDay` source code also existing.)

Reply:
That's a solid idea. I wonder, however, if that wouldn't inadvertently obscure shorter codes that had been introduced ill-advisedly. Thinking out loud -- partner A's code is `orange`, partner B's is `ora`. partner B doesn't know about partner A's code, so they issue a referral-code `orange-juice`, counting on the `ora` part to redirect the client to them. In sorting by the longest first, client is misdirected to partner A instead of B. Sorting by the shortest code first, the opposite problem happens every time partner A's code comes in.

So sorting by longest has the least problems by default, but it doesn't solve for some cases. I think the rule here should be that a partner source code can't be a subset of another partner's code.

* partner A: united-way-